### PR TITLE
[json-rpc] add/delete/toggle timer functionality

### PR DESCRIPTION
**Pull request by stevegal**
updates #8816 which I can't seem to reopen

### CODE REVIEW

File: xbmc/interfaces/json-rpc/PVROperations.cpp [hunk 1]
```diff
@@ -337,6 +337,103 @@ JSONRPC_STATUS CPVROperations::GetTimerDetails(const std::string &method, ITrans
...
+
+  EpgSearchFilter filter;
+  filter.Reset();
+  filter.m_iUniqueBroadcastId = parameterObject["broadcastid"].asUnsignedInteger();
+
+  CFileItemList broadcasts;
+  int resultSize = g_EpgContainer.GetEPGSearch(broadcasts, filter);
+
+  if (resultSize != 1)
+    return InvalidRequest;
```
Comment (Montellese) [inline xbmc/interfaces/json-rpc/PVROperations.cpp]: `InvalidParams`

File: xbmc/interfaces/json-rpc/PVROperations.cpp [hunk 2]
```diff
@@ -337,6 +337,103 @@ JSONRPC_STATUS CPVROperations::GetTimerDetails(const std::string &method, ITrans
...
+  bool repeating = parameterObject["repeating"].asBoolean(false);
+  if (epgTag->HasTimer())
+    return InvalidParams;
+
+  CPVRTimerInfoTagPtr newTimer = CPVRTimerInfoTag::CreateFromEpg(epgTag, repeating);
+  if (newTimer)
+  {
+    bool added = g_PVRTimers->AddTimer(newTimer);
+    if (added)
+    {
```
Comment (Montellese) [inline xbmc/interfaces/json-rpc/PVROperations.cpp]: No curly braces on single line blocks.

File: xbmc/interfaces/json-rpc/PVROperations.cpp [hunk 3]
```diff
@@ -337,6 +337,103 @@ JSONRPC_STATUS CPVROperations::GetTimerDetails(const std::string &method, ITrans
...
+
+  CPVRTimers* timers = g_PVRTimers;
+
+  CPVRTimerInfoTagPtr timer = timers->GetById(parameterObject["timerid"].asInteger());
+  if (!timer)
+    return InvalidParams;
+
+  bool repeating = parameterObject["repeating"].asBoolean(false);
+
+  bool sentOkay = timers->DeleteTimer(timer, false, repeating);
```
Comment (Montellese) [inline xbmc/interfaces/json-rpc/PVROperations.cpp]: `sentOkay`?

File: xbmc/interfaces/json-rpc/PVROperations.cpp [hunk 4]
```diff
@@ -337,6 +337,103 @@ JSONRPC_STATUS CPVROperations::GetTimerDetails(const std::string &method, ITrans
...
+
+  if(sentOkay)
```
Comment (Montellese) [inline xbmc/interfaces/json-rpc/PVROperations.cpp]: Space between `if` and `(`.

File: xbmc/interfaces/json-rpc/PVROperations.cpp [hunk 5]
```diff
@@ -337,6 +337,103 @@ JSONRPC_STATUS CPVROperations::GetTimerDetails(const std::string &method, ITrans
...
+
+  if (resultSize != 1)
+    return InvalidRequest;
+
+  CFileItemPtr broadcast = broadcasts.Get(0);
+
+  if (!broadcast->HasEPGInfoTag())
+    return InvalidRequest;
+
+  CEpgInfoTagPtr epgTag = broadcast->GetEPGInfoTag();
```
Comment (Montellese) [inline xbmc/interfaces/json-rpc/PVROperations.cpp]: The logic from the beginning of this method to this line is identical to the one in `AddTimer()` so maybe you could move it into a separate method which takes a `broadcastid` and returns `CEpgInfoTagPtr`.

File: xbmc/interfaces/json-rpc/PVROperations.cpp [hunk 6]
```diff
@@ -337,6 +337,103 @@ JSONRPC_STATUS CPVROperations::GetTimerDetails(const std::string &method, ITrans
...
+
+  bool repeating = parameterObject["repeating"].asBoolean(false);
+  if (epgTag->HasTimer())
+  {
+    CFileItemPtr timerTag = g_PVRTimers->GetTimerForEpgTag(broadcast.get());
+    g_PVRTimers->DeleteTimer( *timerTag , false, repeating);
```
Comment (Montellese) [inline xbmc/interfaces/json-rpc/PVROperations.cpp]: Shouldn't you check the return type of `DeleteTimer()` and `AddTimer()` (a few lines down)?

File: xbmc/interfaces/json-rpc/schema/methods.json [hunk 7]
```diff
@@ -1895,6 +1895,39 @@
...
     }
   },
+  "PVR.AddTimer": {
+    "type": "method",
+    "description": "Adds a timer",
+    "transport": "Response",
+    "permission": "ControlPVR",
+    "params": [
+      { "name": "broadcastid", "$ref": "Library.Id", "required": true, "description": "the broadcast id of item to record" },
+      { "name": "repeating", "type": "boolean", "required": false, "default": false, "description": "the add should be done for all occurances" }
```
Comment (Montellese) [inline xbmc/interfaces/json-rpc/schema/methods.json]: Specifying `"required": false` is not necessary. That's the default behaviour.

**Comment by Montellese** [comment]
I can't comment on the actual PVR specific stuff but apart from a few potential improvements the JSON-RPC stuff looks ok to me.

File: xbmc/interfaces/json-rpc/PVROperations.cpp (hunk 3 above)
Comment (stevegal) [inline xbmc/interfaces/json-rpc/PVROperations.cpp]: sentOkay because that what's the header file says the functionality is. Just matching that.

**Comment by stevegal** [comment]
just need to be more defensive on the broadcastid search return. Do you want me to rebase this into the "same" commit as 6a37f43?

File: xbmc/interfaces/json-rpc/PVROperations.cpp [hunk 8]
```diff
@@ -337,6 +337,114 @@ JSONRPC_STATUS CPVROperations::GetTimerDetails(const std::string &method, ITrans
...
+  bool repeating = parameterObject["repeating"].asBoolean(false);
+
+  if (epgTag->HasTimer())
+    return InvalidParams;
+
+  CPVRTimerInfoTagPtr newTimer = CPVRTimerInfoTag::CreateFromEpg(epgTag, repeating);
+  if (newTimer)
+  {
+    bool sentOkay = g_PVRTimers->AddTimer(newTimer);
+    if (sentOkay)
```
Comment (Montellese) [inline xbmc/interfaces/json-rpc/PVROperations.cpp]: Actually you don't really need the temporary. Just use
```
if (g_PVRTimers->AddTimer(newTimer))
  return ACK;
```
Same below.

File: xbmc/interfaces/json-rpc/PVROperations.cpp [hunk 9]
```diff
@@ -337,6 +337,114 @@ JSONRPC_STATUS CPVROperations::GetTimerDetails(const std::string &method, ITrans
...
+  CFileItemPtr broadcast;
+  EpgSearchFilter filter;
+  filter.Reset();
+  filter.m_iUniqueBroadcastId = broadcastid;
+
+  CFileItemList broadcasts;
+  int resultSize = g_EpgContainer.GetEPGSearch(broadcasts, filter);
+
+  if (resultSize != 1)
+    return broadcast;
```
Comment (Montellese) [inline xbmc/interfaces/json-rpc/PVROperations.cpp]: `return CFileItemPtr();`

File: xbmc/interfaces/json-rpc/PVROperations.cpp [hunk 10]
```diff
@@ -337,6 +337,114 @@ JSONRPC_STATUS CPVROperations::GetTimerDetails(const std::string &method, ITrans
...
+
+  broadcast = broadcasts.Get(0);
+
+  return broadcast;
```
Comment (Montellese) [inline xbmc/interfaces/json-rpc/PVROperations.cpp]: `return broadcasts.Get(0);` and then you can get rid of the `broadcast` variable.

**Comment by Montellese** [comment]
Force pushing is fine. In the end everything should be squashed into one commit for this PR.

File: xbmc/interfaces/json-rpc/PVROperations.cpp (hunk 9 above)
Comment (stevegal) [inline xbmc/interfaces/json-rpc/PVROperations.cpp]: will the reference above where I've created the shared_ptr not have the same effect?

File: xbmc/interfaces/json-rpc/PVROperations.cpp (hunk 8 above)
Comment (stevegal) [inline xbmc/interfaces/json-rpc/PVROperations.cpp]: Is this kodi style? (I personally don't like a main function hidden in an if statement, but will modify if it's more in line with the rest of style!

File: xbmc/interfaces/json-rpc/PVROperations.cpp (hunk 9 above)
Comment (Montellese) [inline xbmc/interfaces/json-rpc/PVROperations.cpp]: Yes but with the other change I propose further down you can get rid of the fixed reference completely.

File: xbmc/interfaces/json-rpc/PVROperations.cpp (hunk 10 above)
Comment (stevegal) [inline xbmc/interfaces/json-rpc/PVROperations.cpp]: good point!

File: xbmc/interfaces/json-rpc/PVROperations.cpp (hunk 8 above)
Comment (Montellese) [inline xbmc/interfaces/json-rpc/PVROperations.cpp]: If you only need the variable for a single check afterwards yes.

File: xbmc/interfaces/json-rpc/PVROperations.cpp (hunk 9 above)
Comment (stevegal) [inline xbmc/interfaces/json-rpc/PVROperations.cpp]: cool. I'll modify

File: xbmc/interfaces/json-rpc/PVROperations.cpp (hunk 8 above)
Comment (stevegal) [inline xbmc/interfaces/json-rpc/PVROperations.cpp]: cool. Will modify

**Comment by stevegal** [comment]
think that's all in place now.

**Comment by Montellese** [comment]
Yup, much better, thanks. Please squash the two commits into one.

**Comment by stevegal** [comment]
done.

**Comment by stevegal** [comment]
What are the next steps? I'm new to this, so not sure if I should be doing something or not?

**Comment by stefansaraev** [comment]
jenkins build this please

**Comment by stefansaraev** [comment]
@ksooo or any other pvr dev for ack.

**Comment by stevegal** [comment]
Is there anything else I need to do, or should I just be note patient!?

**Comment by Montellese** [comment]
@stevegal We need to wait until someone from the PVR devs has the time to look at this.

File: xbmc/interfaces/json-rpc/PVROperations.cpp [hunk 11]
```diff
@@ -337,6 +337,108 @@ JSONRPC_STATUS CPVROperations::GetTimerDetails(const std::string &method, ITrans
...
+{
+  EpgSearchFilter filter;
+  filter.Reset();
+  filter.m_iUniqueBroadcastId = broadcastid;
+
+  CFileItemList broadcasts;
+  int resultSize = g_EpgContainer.GetEPGSearch(broadcasts, filter);
+
+  if (resultSize != 1)
+    return CFileItemPtr();
```
Comment (Jalle19) [inline xbmc/interfaces/json-rpc/PVROperations.cpp]: Do we not have a better way of indicating "no results found" than this?

Comment (Jalle19) [inline xbmc/interfaces/json-rpc/PVROperations.cpp]: Never mind, saw how it's used a bit further down.

File: xbmc/interfaces/json-rpc/PVROperations.cpp [hunk 12]
```diff
@@ -337,6 +337,108 @@ JSONRPC_STATUS CPVROperations::GetTimerDetails(const std::string &method, ITrans
...
+  return FailedToExecute;
+}
+
+
+JSONRPC_STATUS CPVROperations::DeleteTimer(const std::string &method, ITransportLayer *transport, IClient *client, const CVariant &parameterObject, CVariant &result)
+{
+  if (!g_PVRManager.IsStarted())
+    return FailedToExecute;
+
+  CPVRTimers* timers = g_PVRTimers;
```
Comment (Jalle19) [inline xbmc/interfaces/json-rpc/PVROperations.cpp]: I'd drop this and use `g_PVRTimers` directly instead.

File: xbmc/interfaces/json-rpc/PVROperations.cpp [hunk 13]
```diff
@@ -337,6 +337,108 @@ JSONRPC_STATUS CPVROperations::GetTimerDetails(const std::string &method, ITrans
...
+    return InvalidParams;
+
+  CEpgInfoTagPtr epgTag = broadcast->GetEPGInfoTag();
+
+  if (!epgTag)
+    return InvalidParams;
+
+  bool repeating = parameterObject["repeating"].asBoolean(false);
+  bool sentOkay = false;
+  if (epgTag->HasTimer())
```
Comment (Jalle19) [inline xbmc/interfaces/json-rpc/PVROperations.cpp]: `success` might be a better name here, but it's not a big deal.

File: xbmc/interfaces/json-rpc/PVROperations.cpp [hunk 14]
```diff
@@ -337,6 +337,108 @@ JSONRPC_STATUS CPVROperations::GetTimerDetails(const std::string &method, ITrans
...
+  if (!broadcast->HasEPGInfoTag())
+    return InvalidParams;
+
+  CEpgInfoTagPtr epgTag = broadcast->GetEPGInfoTag();
+
+  if (!epgTag)
+    return InvalidParams;
+
+  bool repeating = parameterObject["repeating"].asBoolean(false);
+  bool sentOkay = false;
```
Comment (Jalle19) [inline xbmc/interfaces/json-rpc/PVROperations.cpp]: `success` might be a better name here, but it's not a big deal.

**Comment by Jalle19** [comment]
Apart from the minor comments this needs an API version bump. See earlier commits for how it's done.

**Comment by Montellese** [comment]
You need to increment the minor version number in `xbmc/interfaces/json-rpc/schema/version.txt` by one and reset the bugfix version number to `0`.

File: xbmc/interfaces/json-rpc/PVROperations.cpp (hunk 12 above)
Comment (opdenkamp) [inline xbmc/interfaces/json-rpc/PVROperations.cpp]: `timers` is used twice. the way it's written now doesn't rely on the compiler optimising this out

**Comment by stevegal** [comment]
Cool. I'll bump that version number and push. I guess I should do a merge of xbmc master into this branch first? Thanks all!

**Comment by stefansaraev** [comment]
>  I guess I should do a merge of xbmc master into this branch first? Thanks all!

something like
```
git remote add upstream https://github.com/xbmc/xbmc.git
git fetch -p --all
git rebase upstream/master
```
should be enough

**Comment by stevegal** [comment]
@stefansaraev @Montellese @Jalle19 @opdenkamp Hope that's all in place now. And thanks for everyones help! Nice and friendly!

File: xbmc/interfaces/json-rpc/PVROperations.cpp (hunk 14 above)
Comment (stevegal) [inline xbmc/interfaces/json-rpc/PVROperations.cpp]: the documentation in the pvr stuff said "sent" that's why I called the variable sentOkay. I've left tis as is, as you say - no big deal

File: xbmc/interfaces/json-rpc/PVROperations.cpp (hunk 12 above)
Comment (stevegal) [inline xbmc/interfaces/json-rpc/PVROperations.cpp]: I've left this as is - thanks

Comment (Jalle19) [inline xbmc/interfaces/json-rpc/PVROperations.cpp]: Okay, guess I missed that

**Comment by Jalle19** [comment]
Thanks

**Pull request merged**